### PR TITLE
Add ability to filter potential demo cards

### PR DESCRIPTION
**Pull request by adamslc**
I think that this is a broadly useful functionality. It is useful for me because I use
`ReTestItems.jl` in a package I maintain, and so I have `_test.jl` files throughout the
repo, including in my examples directory.

### CODE REVIEW

File: src/generate.jl [hunk 1]
```diff
@@ -46,6 +46,8 @@ Processing pipeline:
 * `credit::Bool`: `true` to show a "This page is generated by ..." info. By default it's `true`.
 * `throw_error::Bool`: `true` to throw an error when the julia demo build fails; otherwise it will
   continue the build with warnings.
+* `filter_function::Function`: the function is passed the path of each potential demo card, and the
```
Comment (johnnychen94) [inline src/generate.jl:49]: In addition to the path of the card, an alternative way is to passing the properties of the card (or just the card itself).

File: src/types/section.jl [hunk 2]
```diff
@@ -88,12 +88,12 @@ end
...
 basename(sec::DemoSection) = basename(sec.root)
 
-function DemoSection(root::String)::DemoSection
+function DemoSection(root::String, filter_func=x -> true)::DemoSection
     root = replace(root, r"[/\\]" => Base.Filesystem.path_separator) # windows compatibility
     isdir(root) || throw(ArgumentError("section root should be a valid dir, instead it's $(root)"))
 
     path = joinpath.(root, filter(x->!startswith(x, "."), readdir(root))) # filter out hidden files
-    card_paths = filter(x->isfile(x) && !endswith(x, config_filename), path)
+    card_paths = filter(x->isfile(x) && !endswith(x, config_filename) && filter_func(x), path)
```
Comment (johnnychen94) [inline src/types/section.jl:96]: I think the filtering part is better done in or after R117-R119.

**Review by johnnychen94 — COMMENTED** [review]
I'm okay with having this feature, but I'd like to see a more generic version of it (filtering on properties rather than path).

**Comment by johnnychen94** [comment]
Oops, the new commit seems to introduce some breaking change that I failed to foresee. @adamslc could you help check and fix it?

**Comment by adamslc** [comment]
Whoops, I had a silly bug. Fixed now, and tests pass locally.

**Review by johnnychen94 — APPROVED** [review]

**Pull request merged**